src/compiler/parser.ts

```diff
@@ -1001,10 +1008,12 @@ module ts {
             switch (parsingContext) {
                 case ParsingContext.SourceElements:
                 case ParsingContext.ModuleElements:
-                    return isSourceElement(inErrorRecovery);
+                    // During error recovery we don't treat empty statements as statements
+                    return !(token === SyntaxKind.SemicolonToken && inErrorRecovery) && isModuleElement();
                 case ParsingContext.BlockStatements:
                 case ParsingContext.SwitchClauseStatements:
-                    return isStartOfStatement(inErrorRecovery);
+                    // During error recovery we don't treat empty statements as statements
+                    return !(token === SyntaxKind.SemicolonToken && inErrorRecovery) && isStatement();

```

Comment:
I disagree. We don't propagate this flag to any of the other isXXX helpers and with the current structure all handling of inErrorRecovery is now local to isListElement, making it much clearer what the effect of the flag actually is.